File: examples/week-one-java-examples/src/singletoneager/Singleton.java
package singletoneager;

public class Singleton {
    private static int counter;
    private static Singleton single;

    static {
        counter = 0;
        single = new Singleton();
    }

    private Singleton() {
        counter++;
    }

    public static Singleton getInstance() {
        return single;
    }

    @Override
    public String toString() {
        return getClass() + " " + hashCode() + " " + counter;
    }

    public static void main(String[] args) {
        final int LIMIT = 10;
        System.out.println(Singleton.counter);
        for (int x = 0; x < LIMIT; x++) {
            System.out.println(Singleton.getInstance());
        }
    }
}

File: kotlin-examples-week-one/src/singleton/Singleton.kt
package singleton

object Singleton

class SingletonWithCons private constructor() {

    init {
        counter++
    }

    override fun toString(): String {
        return javaClass.toString() + " " + hashCode() + " " + counter
    }

    companion object {
        var counter = 0
        var instance: SingletonWithCons? = null
            private set

        init {
            instance = SingletonWithCons()
        }
    }
}

fun main(args: Array<String>) {
    val LIMIT = 10
    println(SingletonWithCons.counter)
    for (x in 0 until LIMIT) {
        println(SingletonWithCons.instance)
    }
}